ergotree-interpreter/src/eval/soption.rs

```diff
@@ -0,0 +1,325 @@
+use crate::eval::EvalError;
+use crate::eval::Evaluable;
+
+use ergotree_ir::mir::value::Value;
+
+use super::EvalFn;
+
+pub(crate) static MAP_EVAL_FN: EvalFn = |env, ctx, obj, args| {
+    let input_v = obj;
+    let lambda_v = args
+        .get(0)
+        .cloned()
+        .ok_or_else(|| EvalError::NotFound("map: eval is missing first arg".to_string()))?;
+    let input_v_clone = input_v.clone();
+    let lambda = match &lambda_v {
+        Value::Lambda(l) => Ok(l),
+        _ => Err(EvalError::UnexpectedValue(format!(
+            "expected lambda to be Value::FuncValue got: {0:?}",
+            input_v_clone
+        ))),
+    }?;
+    let mut lambda_call = |arg: Value| {
+        let func_arg = lambda.args.first().ok_or_else(|| {
+            EvalError::NotFound("map: lambda has empty arguments list".to_string())
+        })?;
+        let env1 = env.clone().extend(func_arg.idx, arg);
+        lambda.body.eval(&env1, ctx)
+    };
+    let normalized_input_val: Option<Value> = match input_v {
+        Value::Opt(opt) => Ok(*opt),
+        _ => Err(EvalError::UnexpectedValue(format!(
+            "expected map input to be Value::Opt, got: {0:?}",
+            input_v
+        ))),
+    }?;
+
+    match normalized_input_val {
+        Some(t) => Ok(Value::Opt(Box::new(lambda_call(t)?.into()))),
+        _ => Ok(Value::Opt(Box::new(None))),
+    }
+};
+
+pub(crate) static FILTER_EVAL_FN: EvalFn = |env, ctx, obj, args| {
+    let input_v = obj;
+    let lambda_v = args
+        .get(0)
+        .cloned()
+        .ok_or_else(|| EvalError::NotFound("filter: eval is missing first arg".to_string()))?;
+    let input_v_clone = input_v.clone();
+    let lambda = match &lambda_v {
+        Value::Lambda(l) => Ok(l),
+        _ => Err(EvalError::UnexpectedValue(format!(
+            "expected lambda to be Value::FuncValue got: {0:?}",
+            input_v_clone
+        ))),
+    }?;
+    let mut predicate_call = |arg: Value| {
+        let func_arg = lambda.args.first().ok_or_else(|| {
+            EvalError::NotFound("filter: lambda has empty arguments list".to_string())
+        })?;
+        let env1 = env.clone().extend(func_arg.idx, arg);
+        lambda.body.eval(&env1, ctx)
+    };
+    let normalized_input_val: Option<Value> = match input_v {
+        Value::Opt(opt) => Ok(*opt),
+        _ => Err(EvalError::UnexpectedValue(format!(
+            "expected filter input to be Value::Opt, got: {0:?}",
+            input_v
+        ))),
+    }?;
+
+    if let Some(v) = normalized_input_val {
+        if let Value::Boolean(predicate_res) = predicate_call(v.clone())? {
```

Comment:
I believe if `predicate_call` return something other than `Value::Boolean` we silently return `None`. I'd rather return an error in this case.

Reply:
So I figured that this case would be handled by `MethodCall::new` where it checks expected types according to the `SMethodDesc`.  In a unit test, I am unable to even instantiate the `MethodCall` if the predicate's `Expr.tpe()` is not `SType::SBoolean`.

In any case, I have added an explicit check for this in the eval function as well.  I switched to match statements to make it more readable.

Reply:
Type check on IR node construction checks the declared type while on the evaluation we have a resulted value that is not guaranteed to have that declared type. Think of not-yet-evaluated expression as an opaque box with a sticker(type) on it. Only when it is opened (expression is evaluated) we can see the result value and its type. Due to possible errors on evaluation, compilation, serialization, etc. I prefer to "double" check that what we got is the same that what was expected (sticker) before propagating an invalid value further up the expression tree.

Reply:
Makes sense, thanks for the explanation.